### ISSUE TITLE
Many props undefined when getting directory contents from NextCloud webdav

### ISSUE THREAD

**Issue by Dogcatfee** [perry-mitchell/webdav-client#194]
Webdav: 3.0.0
NextCloud: 16.0.1

As shown in the code below, I have commented out most of this function to get directory listing in NextCloud. I'm not really sure if this is a NextCloud issue or not.

``` bash
 15 function propsToStat(props, filename, isDetailed = false) {
 16   const path = require("path-posix"); // Last modified time, raw size, item type and mime
 17 
 18 
 19   //const lastMod = props.getlastmodified,
 20   //      _props$getcontentleng = props.getcontentlength,
 21   //      rawSize = _props$getcontentleng === void 0 ? "0" : _props$getcontentleng,
 22   //      resourceType = props.resourcetype,
 23   //      mimeType = props.getcontenttype,
 24   //      etag = props.getetag;
 25   //const type = resourceType && typeof resourceType === "object" && typeof resourceType.collection !== "undefined" ? "directory" : "file";
 26   const stat = {
 27     filename: filename,
 28     basename: path.basename(filename),
 29     //lastmod: lastMod,
 30     //size: parseInt(rawSize, 10),
 31     //type: type,
 32     //etag: typeof etag === "string" ? etag.replace(/"/g, "") : null
 33   };
 34 
 35   //if (type === "file") {
 36   //  stat.mime = mimeType && typeof mimeType === "string" ? mimeType.split(";")[0] : "";
 37   //}
 38 
 39   if (isDetailed) {
 40     stat.props = props;
 41   }
 42 
 43   return stat;
 44 }
```

**Comment by perry-mitchell** [comment]
Any way you could share some of the XML output from your Nextcloud? The tests are passing without change so I don’t believe it’s the upgrade. Did you notice this issue with an earlier version of WebDAV client or have you just started using it?

**Comment by Dogcatfee** [comment]
Moved stuff into Documents folder to cut down on XML size and identifiable info, uncommented the lines and tested again with node to make sure that it still crashes. Looking through it I see a bunch of 404 entries that don't have the needed props, this is probably what was causing the crash. 
``` curl -X PROPFIND -u 'user:pass' 'https://server.com/remove.php/webdav'```

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:nc="http://nextcloud.org/ns" xmlns:oc="http://owncloud.org/ns" xmlns:s="http://sabredav.org/ns">
   <d:response>
      <d:href>/remote.php/webdav/</d:href>
      <d:propstat>
         <d:prop>
            <d:getlastmodified>Sun, 02 Feb 2020 08:59:02 GMT</d:getlastmodified>
            <d:resourcetype>
               <d:collection />
            </d:resourcetype>
            <d:quota-used-bytes>681578939</d:quota-used-bytes>
            <d:quota-available-bytes>-3</d:quota-available-bytes>
            <d:getetag>"5e368f56acbeb"</d:getetag>
         </d:prop>
         <d:status>HTTP/1.1 200 OK</d:status>
      </d:propstat>
   </d:response>
   <d:response>
      <d:href>/remote.php/webdav/.gitignore</d:href>
      <d:propstat>
         <d:prop>
            <d:getlastmodified>Fri, 15 Nov 2019 14:51:23 GMT</d:getlastmodified>
            <d:getcontentlength>25</d:getcontentlength>
            <d:resourcetype />
            <d:getetag>"6e5386d99de14eb9f8f557728d1a91a6"</d:getetag>
            <d:getcontenttype>application/octet-stream</d:getcontenttype>
         </d:prop>
         <d:status>HTTP/1.1 200 OK</d:status>
      </d:propstat>
      <d:propstat>
         <d:prop>
            <d:quota-used-bytes />
            <d:quota-available-bytes />
         </d:prop>
         <d:status>HTTP/1.1 404 Not Found</d:status>
      </d:propstat>
   </d:response>
   <d:response>
      <d:href>/remote.php/webdav/.stack-work/</d:href>
      <d:propstat>
         <d:prop>
            <d:getlastmodified>Mon, 18 Nov 2019 04:35:40 GMT</d:getlastmodified>
            <d:resourcetype>
               <d:collection />
            </d:resourcetype>
            <d:quota-used-bytes>0</d:quota-used-bytes>
            <d:quota-available-bytes>-3</d:quota-available-bytes>
            <d:getetag>"5dd21f9c501ae"</d:getetag>
         </d:prop>
         <d:status>HTTP/1.1 200 OK</d:status>
      </d:propstat>
      <d:propstat>
         <d:prop>
            <d:getcontentlength />
            <d:getcontenttype />
         </d:prop>
         <d:status>HTTP/1.1 404 Not Found</d:status>
      </d:propstat>
   </d:response>
   <d:response>
      <d:href>/remote.php/webdav/Documents/</d:href>
      <d:propstat>
         <d:prop>
            <d:getlastmodified>Sun, 02 Feb 2020 08:59:02 GMT</d:getlastmodified>
            <d:resourcetype>
               <d:collection />
            </d:resourcetype>
            <d:quota-used-bytes>325687491</d:quota-used-bytes>
            <d:quota-available-bytes>-3</d:quota-available-bytes>
            <d:getetag>"5e368f56acbeb"</d:getetag>
         </d:prop>
         <d:status>HTTP/1.1 200 OK</d:status>
      </d:propstat>
      <d:propstat>
         <d:prop>
            <d:getcontentlength />
            <d:getcontenttype />
         </d:prop>
         <d:status>HTTP/1.1 404 Not Found</d:status>
      </d:propstat>
   </d:response>
   <d:response>
      <d:href>/remote.php/webdav/Pictures/</d:href>
      <d:propstat>
         <d:prop>
            <d:getlastmodified>Sun, 17 Nov 2019 06:50:52 GMT</d:getlastmodified>
            <d:resourcetype>
               <d:collection />
            </d:resourcetype>
            <d:quota-used-bytes>328268893</d:quota-used-bytes>
            <d:quota-available-bytes>-3</d:quota-available-bytes>
            <d:getetag>"5dd0edcd7fb74"</d:getetag>
         </d:prop>
         <d:status>HTTP/1.1 200 OK</d:status>
      </d:propstat>
      <d:propstat>
         <d:prop>
            <d:getcontentlength />
            <d:getcontenttype />
         </d:prop>
         <d:status>HTTP/1.1 404 Not Found</d:status>
      </d:propstat>
   </d:response>
   <d:response>
      <d:href>/remote.php/webdav/Shared/</d:href>
      <d:propstat>
         <d:prop>
            <d:getlastmodified>Wed, 11 Dec 2019 09:29:04 GMT</d:getlastmodified>
            <d:resourcetype>
               <d:collection />
            </d:resourcetype>
            <d:quota-used-bytes>27622530</d:quota-used-bytes>
            <d:quota-available-bytes>-3</d:quota-available-bytes>
            <d:getetag>"5df0b6e0a7430"</d:getetag>
         </d:prop>
         <d:status>HTTP/1.1 200 OK</d:status>
      </d:propstat>
      <d:propstat>
         <d:prop>
            <d:getcontentlength />
            <d:getcontenttype />
         </d:prop>
         <d:status>HTTP/1.1 404 Not Found</d:status>
      </d:propstat>
   </d:response>
</d:multistatus>
```

**Comment by Dogcatfee** [comment]
Thanks for the XML idea, I had no idea that was a thing. This doesn't crash, but it feels bad to use so many ternary operators.

``` javascript
 15 function propsToStat(props, filename, isDetailed = false) {
 16   const path = require("path-posix"); // Last modified time, raw size, item type and mime
 17 
 18   const lastMod = props ? props.getlastmodified : null,
 19         _props$getcontentleng = props ? props.getcontentlength: null,
 20         rawSize = props ? _props$getcontentleng === void 0 ? "0" : _props$getcontentleng : null,
 21         resourceType = props ? props.resourcetype : null,
 22         mimeType = props ? props.getcontenttype : null,
 23         etag = props ? props.getetag : null;
 24   const type = resourceType && typeof resourceType === "object" && typeof resourceType.collection !== "undefined" ? "directory" : "file";
 25   const stat = {
 26     filename: filename,
 27     basename: path.basename(filename),
 28     lastmod: lastMod,
 29     size: parseInt(rawSize, 10),
 30     type: type,
 31     etag: typeof etag === "string" ? etag.replace(/"/g, "") : null
 32   };
 33 
 34   if (type === "file") {
 35     stat.mime = mimeType && typeof mimeType === "string" ? mimeType.split(";")[0] : "";
 36   }
 37 
 38   if (props && isDetailed) {
 39     stat.props = props;
 40   }
 41 
 42   return stat;
 43 }
 44 
```

**Comment by skjnldsv** [comment]
Can confirm!
parseXML returns undefined

**Comment by skjnldsv** [comment]
``` xml
<?xml version="1.0" encoding="UTF-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:nc="http://nextcloud.org/ns" xmlns:oc="http://owncloud.org/ns" xmlns:s="http://sabredav.org/ns">
	<d:response>
		<d:href>/remote.php/dav/files/admin/Photos/daniel-apodaca-NwCIEX2UIL0-unsplash.jpg</d:href>
		<d:propstat>
			<d:prop>
				<d:getlastmodified>Fri, 25 Oct 2019 08:56:23 GMT</d:getlastmodified>
				<d:getetag>"c2b3a667cb562d19fbb9c7d0b500cb9f"</d:getetag>
				<d:getcontenttype>image/jpeg</d:getcontenttype>
				<d:resourcetype />
				<oc:fileid>14445</oc:fileid>
				<oc:permissions>RGDNVW</oc:permissions>
				<oc:size>1351471</oc:size>
				<d:getcontentlength>1351471</d:getcontentlength>
				<nc:has-preview>true</nc:has-preview>
				<nc:mount-type />
				<x1:share-permissions xmlns:x1="http://open-collaboration-services.org/ns">19</x1:share-permissions>
				<oc:tags />
				<oc:favorite>0</oc:favorite>
				<oc:comments-unread>0</oc:comments-unread>
				<oc:owner-id>admin</oc:owner-id>
				<oc:owner-display-name>Administrator</oc:owner-display-name>
				<oc:share-types />
			</d:prop>
			<d:status>HTTP/1.1 200 OK</d:status>
		</d:propstat>
		<d:propstat>
			<d:prop>
				<nc:is-encrypted />
			</d:prop>
			<d:status>HTTP/1.1 404 Not Found</d:status>
		</d:propstat>
	</d:response>
</d:multistatus>
```

**Comment by perry-mitchell** [comment]
@Dogcatfee Your response from Nextcloud has a lot of empty properties, like so:

```xml
<d:prop>
  <d:getcontentlength />
  <d:getcontenttype />
</d:prop>
```

So that would definitely cause an issue in this case. We'll need to handle these in a smarter way.

@skjnldsv Your XML looks a bit more complete, so I can't immediately see why it's failing. I'll jump on it shortly to see where the parsing dies. Thanks for the help!

**Comment by perry-mitchell** [comment]
@skjnldsv Weird, I can parse your XML just fine using `parseXML`..

```json
{
  "multistatus": {
    "response": {
      "href": "/remote.php/dav/files/admin/Photos/daniel-apodaca-NwCIEX2UIL0-unsplash.jpg",
      "propstat": [
        {
          "prop": {
            "getlastmodified": "Fri, 25 Oct 2019 08:56:23 GMT",
            "getetag": "\"c2b3a667cb562d19fbb9c7d0b500cb9f\"",
            "getcontenttype": "image/jpeg",
            "resourcetype": "",
            "fileid": 14445,
            "permissions": "RGDNVW",
            "size": 1351471,
            "getcontentlength": 1351471,
            "has-preview": true,
            "mount-type": "",
            "share-permissions": 19,
            "tags": "",
            "favorite": 0,
            "comments-unread": 0,
            "owner-id": "admin",
            "owner-display-name": "Administrator",
            "share-types": ""
          },
          "status": "HTTP/1.1 200 OK"
        },
        {
          "prop": {
            "is-encrypted": ""
          },
          "status": "HTTP/1.1 404 Not Found"
        }
      ]
    }
  }
}
```

**Comment by perry-mitchell** [comment]
Connecting to a demo Nextcloud account, I'm seeing the props parsing issue, but that's because nextcloud is doing something weird with the props:

```xml
<d:response>
    <d:href>/remote.php/dav/files/piyaA6MMjrSRBQdC/Documents/</d:href>
    <d:propstat>
        <d:prop>
            <d:getlastmodified>Tue, 04 Feb 2020 18:10:04 GMT</d:getlastmodified>
            <d:resourcetype>
                <d:collection />
            </d:resourcetype>
            <d:quota-used-bytes>405265</d:quota-used-bytes>
            <d:quota-available-bytes>-3</d:quota-available-bytes>
            <d:getetag>&quot;5e39b37cc14a5&quot;</d:getetag>
        </d:prop>
        <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
    <d:propstat>
        <d:prop>
            <d:getcontentlength />
            <d:getcontenttype />
        </d:prop>
        <d:status>HTTP/1.1 404 Not Found</d:status>
    </d:propstat>
</d:response>
```

Notice how `propstat` is an ARRAY here.. so this parses to:

```json
{
  "href": "/remote.php/dav/files/piyaA6MMjrSRBQdC/Documents/",
  "propstat": [
    {
      "prop": {
        "getlastmodified": "Tue, 04 Feb 2020 18:10:04 GMT",
        "resourcetype": {
          "collection": ""
        },
        "quota-used-bytes": 405265,
        "quota-available-bytes": -3,
        "getetag": "&quot;5e39b37cc14a5&quot;"
      },
      "status": "HTTP/1.1 200 OK"
    },
    {
      "prop": {
        "getcontentlength": "",
        "getcontenttype": ""
      },
      "status": "HTTP/1.1 404 Not Found"
    }
  ]
}
```

Which breaks webdav-client because it's expecting `propstat` to be an object with `prop` inside of it.

**Comment by perry-mitchell** [comment]
Please see the referenced PR.

**Comment by perry-mitchell** [comment]
Released in 3.1.0
